parlai/core/torch_agent.py

```diff
@@ -366,60 +442,67 @@ def observe(self, observation):
         return self.batch_act([self.observation])[0]
 
     def batch_act(self, observations):
+        """Process a batch of observations (batchsize list of message dicts).
+
+        These observations have been preprocessed by the observe method.
+
+        Subclasses can override this for special functionality, but if the
+        default behaviors are fine then just override the ``train_step`` and
+        ``eval_step`` methods instead. The former is called when labels are
+        present in the observations batch; otherwise, the latter is called.
+        """
         batch_size = len(observations)
-        # initialize a table of replies with this agent's id
+        # initialize a list of replies with this agent's id
         batch_reply = [{'id': self.getID()} for _ in range(batch_size)]
 
+        # check if there are any labels available, if so we will train on them
         is_training = any(['labels' in obs for obs in observations])
 
+        # convert the observations into vectors
         vec_obs = [self.vectorize(obs) for obs in observations]
 
+        # create a batch from the vectors
         batch = self.batchify(vec_obs)
 
-        if batch.text_vec is None:
-            return batch_reply
-
         if is_training:
-            output = self.train_step(batch.text_vec, batch.label_vec,
-                                     batch.candidates)
+            output = self.train_step(batch)
         else:
-            output = self.eval_step(batch.text_vec, batch.label_vec,
-                                    batch.candidates)
+            output = self.eval_step(batch)
 
-        if isinstance(output, tuple):
-            predictions = output[0]
-            candidate_preds = output[1]
-        else:
-            predictions = output
-            candidate_preds = None
+        if output is None:
+            return batch_reply
 
-        self.match_batch(batch_reply, batch.valid_indices,
-                         predictions=predictions,
-                         candidate_preds=candidate_preds)
+        self.match_batch(batch_reply, batch.valid_indices, output)
 
         return batch_reply
 
-    def train_step(self, xs, ys=None, cands=None, *args, **kwargs):
+    def train_step(self, batch):
+        """Process one batch with training labels."""
         raise NotImplementedError('Abstract class: user must implement train_step')
 
-    def eval_step(self, xs, ys=None, cands=None, *args, **kwargs):
+    def eval_step(self, batch):
+        """Process one batch but do not train on it."""
         raise NotImplementedError('Abstract class: user must implement eval_step')
 
 
 class Beam(object):
-    def __init__(self, beam_size, min_length=3, padding_token=0, bos_token=1, eos_token=2, min_n_best=3, cuda='cpu'):
-        """
-        Generic beam class. It keeps information about beam_size hypothesis.
+    """Generic beam class. It keeps information about beam_size hypothesis."""
```

Comment:
IMHO these formatting changes/doc/comment changes for `Beam` should be in a different PR.